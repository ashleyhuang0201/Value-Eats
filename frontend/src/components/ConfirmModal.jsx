import React from "react";
import {
  Dialog,
  DialogContent,
  DialogActions,
  Box,
  Slide,
} from "@material-ui/core";
import { DialogTitleStyled } from "../styles/DialogTitleStyled";
import { ModalButton } from "../styles/ModalButton";
import { CloseIconStyled } from "../styles/CloseIconStyled";
import CloseIcon from "@material-ui/icons/Close";

const Transition = React.forwardRef(function Transition (props, ref) {
  return <Slide direction="up" ref={ref} {...props} />;
});

export default function ConfirmModal ({
  open,
  handleClose,
  title,
  message,
  handleConfirm, // handles text for right button
  confirmText = "Confirm", // handles the function for the right button
  handleDeny = null, // handles text for the left button
  denyText = "Cancel", // handles the function for the left button
}) {
  return (
    <Dialog
      onClose={handleClose}
      aria-labelledby="customized-dialog-title"
      open={open}
      TransitionComponent={Transition}
      keepMounted
    >
      <DialogTitleStyled id="customized-dialog-title" onClose={handleClose}>
        {title}
      </DialogTitleStyled>
      <CloseIconStyled aria-label="close" onClick={handleClose}>
        <CloseIcon />
      </CloseIconStyled>
      <DialogContent dividers>{message}</DialogContent>
      <DialogActions>
        <Box>
          <ModalButton
            autoFocus
            onClick={() => {
              if (handleDeny) {
                handleDeny();
              }
              handleClose();
            }}
            color="primary"
          >
            {denyText}
          </ModalButton>
          <ModalButton autoFocus onClick={handleConfirm} color="primary">
            {confirmText}
          </ModalButton>
        </Box>
      </DialogActions>
    </Dialog>
  );
}
