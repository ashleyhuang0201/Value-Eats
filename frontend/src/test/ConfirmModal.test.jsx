import { shallow } from "enzyme";
import React from "react";
import ConfirmModal from "../components/ConfirmModal";
import EnzymeToJson from "enzyme-to-json";

describe("ConfirmModal", () => {
  const noop = () => {};
  it("can work in the base case", () => {
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
    />);
    expect(confirmWrapper.exists()).toBeTruthy();
  });

  it("handles default title and message are passed in correctly", () => {
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
    />);
    expect(confirmWrapper.text().includes("Test title")).toBe(true);
    expect(confirmWrapper.text().includes("Test description")).toBe(true);
    expect(confirmWrapper.text().includes("Test description that is not in")).toBe(false);
  });

  it("handles default confirm text correctly", () => {
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
    />);
    expect(confirmWrapper.text().includes("Confirm")).toBe(true);
  });

  it("handles default cancel text correctly", () => {
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
    />);
    expect(confirmWrapper.text().includes("Cancel")).toBe(true);
  });

  it("handles custom confirm text correctly", () => {
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
      confirmText={"This will confirm"}
    />);
    expect(confirmWrapper.text().includes("This will confirm")).toBe(true);
    expect(confirmWrapper.text().includes("Confirm")).toBe(false);
  });

  it("handles custom cancel text correctly", () => {
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
      denyText={"This will cancel"}
    />);
    expect(confirmWrapper.text().includes("This will cancel")).toBe(true);
    expect(confirmWrapper.text().includes("Cancel")).toBe(false);
  });

  it("handles close modal correctly", () => {
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
      denyText={"This will cancel"}
    />);
    expect(confirmWrapper.find("ConfirmModal").exists()).toBeFalsy();
  });

  it("handles close correctly when hitting close button in top right", () => {
    const onClick = jest.fn();
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={onClick}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
      denyText={"This will cancel"}
    />);
    confirmWrapper.simulate("close");
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("handles cancel - default behaviour", () => {
    const onClick = jest.fn();
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={onClick}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
      denyText={"This will cancel"}
    />);
    const buttons = confirmWrapper.children().at(3).children().children();
    buttons.at(0).simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("handles cancel - custom behaviour", () => {
    const onClick = jest.fn();
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
      handleDeny={onClick}
      denyText={"This will cancel"}
    />);
    const buttons = confirmWrapper.children().at(3).children().children();
    buttons.at(0).simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it("handles confirm", () => {
    const onClick = jest.fn();
    const confirmWrapper = shallow(<ConfirmModal open={true}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={onClick}
      denyText={"This will cancel"}
    />);
    const buttons = confirmWrapper.children().at(3).children().children();
    buttons.at(1).simulate("click");
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  /* Snapshot tests - creates snapshots for new tests, otherwise will check the existing tests. To update existing tests,
  follow the prompts on the screen after running test. Done to avoid regression. */

  it("renders with minimal props", () => {
    const confirmWrapper = shallow(<ConfirmModal handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
    />);
    expect(EnzymeToJson(confirmWrapper)).toMatchSnapshot();
  });

  it("renders with custom confirm and deny text", () => {
    const confirmWrapper = shallow(<ConfirmModal handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
      confirmText={"Continue"}
      denyText={"Stop"}
    />);
    expect(EnzymeToJson(confirmWrapper)).toMatchSnapshot();
  });

  it("renders correctly when open is false", () => {
    const confirmWrapper = shallow(<ConfirmModal open={false}
      handleClose={noop}
      title={"Test title"}
      message={"Test description"}
      handleConfirm={noop}
    />);
    expect(EnzymeToJson(confirmWrapper)).toMatchSnapshot();
  });
});
