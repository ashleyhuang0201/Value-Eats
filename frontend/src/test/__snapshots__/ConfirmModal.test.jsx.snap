// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal renders correctly when open is false 1`] = `null`;

exports[`ConfirmModal renders with custom confirm and deny text 1`] = `null`;

exports[`ConfirmModal renders with minimal props 1`] = `null`;
